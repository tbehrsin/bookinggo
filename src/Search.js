import React, { useState, useEffect } from 'react';

export const url = (count, terms) => `https://www.rentalcars.com/FTSAutocomplete.do?solrIndex=fts_en&solrRows=${count}&solrTerm=${terms}`;
export const placeholderText = 'city, airport, station, region and district...';

const Search = () => {
  const [searchTerms, setSearchTerms] = useState('');
  const [results, setResults] = useState([]);

  const onChange = (event) => {
    setSearchTerms(event.target.value);
  };

  useEffect(() => {
    let cancel = false;

    if (searchTerms.trim().length < 2) {
      setResults([]);
    } else {
      (async () => {
        const response = await fetch(url(6, searchTerms.trim()));
        const json = await response.json();
        if (!cancel) {
          setResults(json.results.docs);
        }
      })().catch(err => console.error(err));
    }

    return () => {
      cancel = true;
    };
  }, [searchTerms]);

  return (
    <>
      <label htmlFor="search">
        <p>Pick-up Location</p>
        <div className="Search-input">
          <input type="text" id="search" autoFocus placeholder={placeholderText} onChange={onChange} value={searchTerms} />
          {!!results.length && (
            <ol className="Search-results">
              {results.map((doc, i) => (
                <li key={`${i}`}>{doc.name}</li>
              ))}
            </ol>
          )}
        </div>
      </label>
    </>
  );
}

export default Search;
