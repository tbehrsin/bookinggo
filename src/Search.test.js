import React from 'react';
import { create, act } from 'react-test-renderer';

import Search, { url } from './Search';

describe('Search', () => {
  const immediate = () => new Promise((resolve) => setImmediate(resolve));

  const mockJsonPromise = Promise.resolve({
    results: {
      docs: [
        { name: 'Manchester Airport'},
        { name: 'Manchester' },
        { name: 'Manchester - Piccadilly Train Station'},
        { name: 'Ronaldsway Airport' },
        { name: 'Oldham' },
        { name: 'Bolton' }
      ]
    }
  });

  const mockFetchPromise = Promise.resolve({
    json: () => mockJsonPromise
  });

  const mockFetchPromiseNoResults = Promise.resolve({
    json: () => Promise.resolve({
      results: {
        docs: [
          { name: 'No results found' }
        ]
      }
    })
  });

  test('matches snapshot', () => {
    let renderer;
    act(() => {
      renderer = create(<Search />);
    });
    const output = renderer.toJSON();
    expect(output).toMatchSnapshot();
  });

  test('does not search with 1 character', () => {
    jest.spyOn(global, 'fetch');

    let renderer;
    act(() => {
      renderer = create(<Search />);
    });
    const instance = renderer.root;
    const input = instance.findByType('input');

    act(() => {
      input.props.onChange({ target: { value: 'M' } });
    });

    expect(global.fetch).toHaveBeenCalledTimes(0);

    expect(renderer.toJSON()).toMatchSnapshot();

    global.fetch.mockClear();
  });

  test('searches with 2 or more characters and displays 6 results', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);

    let renderer;
    act(() => {
      renderer = create(<Search />);
    });
    const instance = renderer.root;
    const input = instance.findByType('input');

    act(() => {
      input.props.onChange({ target: { value: 'Ma' } });
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(url(6, 'Ma'));

    await act(async () => {
      await immediate();
    });

    expect(renderer.toJSON()).toMatchSnapshot();

    global.fetch.mockClear();
  });

  test('displays "No results found"', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromiseNoResults);

    let renderer;
    act(() => {
      renderer = create(<Search />);
    });
    const instance = renderer.root;
    const input = instance.findByType('input');

    act(() => {
      input.props.onChange({ target: { value: 'xgdgfgd' } });
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(url(6, 'xgdgfgd'));

    await act(async () => {
      await immediate();
    });

    expect(renderer.toJSON()).toMatchSnapshot();

    global.fetch.mockClear();
  });

  test('hides results box', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);

    let renderer;
    act(() => {
      renderer = create(<Search />);
    });
    const instance = renderer.root;
    const input = instance.findByType('input');

    act(() => {
      input.props.onChange({ target: { value: 'Ma' } });
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(url(6, 'Ma'));

    await act(async () => {
      await immediate();
    });

    act(() => {
      input.props.onChange({ target: { value: 'M' } });
    })

    expect(global.fetch).toHaveBeenCalledTimes(1);

    expect(renderer.toJSON()).toMatchSnapshot();

    global.fetch.mockClear();
  });

  test('cancels fetch promise', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);

    let renderer;
    act(() => {
      renderer = create(<Search />);
    });
    const instance = renderer.root;
    const input = instance.findByType('input');

    act(() => {
      input.props.onChange({ target: { value: 'Ma' } });
    });

    act(() => {
      input.props.onChange({ target: { value: 'M' } });
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(url(6, 'Ma'));

    await act(async () => {
      await immediate();
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);

    expect(renderer.toJSON()).toMatchSnapshot();

    global.fetch.mockClear();
  });
});
