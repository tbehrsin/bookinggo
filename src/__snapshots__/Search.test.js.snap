// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search cancels fetch promise 1`] = `
<label
  htmlFor="search"
>
  <p>
    Pick-up Location
  </p>
  <div
    className="Search-input"
  >
    <input
      autoFocus={true}
      id="search"
      onChange={[Function]}
      placeholder="city, airport, station, region and district..."
      type="text"
      value="M"
    />
  </div>
</label>
`;

exports[`Search displays "No results found" 1`] = `
<label
  htmlFor="search"
>
  <p>
    Pick-up Location
  </p>
  <div
    className="Search-input"
  >
    <input
      autoFocus={true}
      id="search"
      onChange={[Function]}
      placeholder="city, airport, station, region and district..."
      type="text"
      value="xgdgfgd"
    />
    <ol
      className="Search-results"
    >
      <li>
        No results found
      </li>
    </ol>
  </div>
</label>
`;

exports[`Search does not search with 1 character 1`] = `
<label
  htmlFor="search"
>
  <p>
    Pick-up Location
  </p>
  <div
    className="Search-input"
  >
    <input
      autoFocus={true}
      id="search"
      onChange={[Function]}
      placeholder="city, airport, station, region and district..."
      type="text"
      value="M"
    />
  </div>
</label>
`;

exports[`Search hides results box 1`] = `
<label
  htmlFor="search"
>
  <p>
    Pick-up Location
  </p>
  <div
    className="Search-input"
  >
    <input
      autoFocus={true}
      id="search"
      onChange={[Function]}
      placeholder="city, airport, station, region and district..."
      type="text"
      value="M"
    />
  </div>
</label>
`;

exports[`Search matches snapshot 1`] = `
<label
  htmlFor="search"
>
  <p>
    Pick-up Location
  </p>
  <div
    className="Search-input"
  >
    <input
      autoFocus={true}
      id="search"
      onChange={[Function]}
      placeholder="city, airport, station, region and district..."
      type="text"
      value=""
    />
  </div>
</label>
`;

exports[`Search searches with 2 or more characters and displays 6 results 1`] = `
<label
  htmlFor="search"
>
  <p>
    Pick-up Location
  </p>
  <div
    className="Search-input"
  >
    <input
      autoFocus={true}
      id="search"
      onChange={[Function]}
      placeholder="city, airport, station, region and district..."
      type="text"
      value="Ma"
    />
    <ol
      className="Search-results"
    >
      <li>
        Manchester Airport
      </li>
      <li>
        Manchester
      </li>
      <li>
        Manchester - Piccadilly Train Station
      </li>
      <li>
        Ronaldsway Airport
      </li>
      <li>
        Oldham
      </li>
      <li>
        Bolton
      </li>
    </ol>
  </div>
</label>
`;
